### PR TITLE
report busy indicator

### LINKED ISSUE #10644: Indeterminate Progress bars reported as "half-checked"
### Steps to reproduce:
Open https://codepen.io/jnurthen/full/qBErmdB
The page has a link followed by 2 progress bars.
The first is Labelled Loading.... and the Valuetext is set to 50%
The second is also Labelled Loading but has no valuetext (or valuenow)

### Actual behavior:
NVDA reads the indeterminate progressbar as "half-checked":
Speech: `progress bar  half checked    0`
Braille: `⣏⣸⣹ prgbar 0 frame end`

### Expected behavior:
A better term than half-checked should be used for an indeterminate progress bar

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
2019.2.1

#### Windows version:
Windows 10 19013.1122

#### Name and version of other software in use when reproducing the issue:
FF 71.0
Chrome 79.0.3945.88

### PR DESCRIPTION
**Pull request by feerrenrut**
### Link to issue number:
fixes #10644

### Summary of the issue:
On the web, widgets with `role="progressbar"` and no `aria-valuenow` are reported by NVDA as "progress bar  half checked  zero".
These widgets are exposed by the browser with:
- IAccessible accRole `ROLE_SYSTEM_PROGRESSBAR`
- IAccessible accState `STATE_SYSTEM_MIXED, STATE_SYSTEM_INDETERMINATE`.

In chrome the value is always "0" in this case.
Note, in firefox, if `aria-valuetext` is provided, `STATE_SYSTEM_INDETERMINATE` is not set and `accValue` has the `aria-valuetext`.

Test cases on code pen [editor](https://codepen.io/reefturner/pen/KKyoYJb?editors=1100), [full](https://codepen.io/reefturner/full/KKyoYJb)

### Description of how this pull request fixes the issue:
An 'indeterminate' progress bar has no value attribute, it can not convey progress, only activity.
Rather than reporting as a progress bar it is given a new NVDA role of "busy indicator" and its value ("zero") is no longer reported.

Technical details.
- Code to convert decode the bit flags (integer) supplied by IA / IA2 into a set of NVDA States was repeated, this has been extracted to a   fucntion, centralized in IAccessibleHandler, and optimised.
  Commits:
  - Extract states calculations
  - Optimise / simplify state calculations
- The initial approach was to report "progress bar indeterminate", see commit: Report as 'progress bar indeterminate'

Value was exposed as the text node in the virtual buffer, in this case space is now used.

### Testing strategy:
Test cases on code pen [editor](https://codepen.io/reefturner/pen/KKyoYJb?editors=1100), [full](https://codepen.io/reefturner/full/KKyoYJb)
- Tested chrome and firefox, speech and braille
- Focus mode (navigate with tab)
- Browse mode (navigate with tab)
- Browse mode (navigate with down arrow)

### Known issues with pull request:
None

### Change log entries:
New features
Changes
```
- Added new role for "busy indicator" controls. (#10644)
```

### Code Review Checklist:
- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

**Comment by CyrilleB79** [comment]
I am afraid this new role can create confusion. Is there a busy indicator object in one of the various interface specifications (IAccessible, UIA, etc.)
More generally, if I speak of a busy indicator, does someone (sighted or not) know what it is?

From a user point of view, busy indicator may let think to the unrelated following concepts:
- The browser indicates "busy" when the page is loading
- "busy" may also refer to a mouse pointer shape (e.g. hourglass, clock, etc.) indicating that the computer is busy.

And overall, we lose the information that the "busy indicator" is actually a progress bar. In case of collaboration with a sighted person, this creates communication difficulties.

I have not followed all the discussion but could the following solutions be implemented instead?
Give the progress bar with undetermined stat a state or value of "running", "working", etc.? I agree that it's not so easy to find a correct wording for it. But creating a new role instead seems more confusing to me.

**Review by Qchristensen — APPROVED** [review]
bsyind for Braille looks good.

File: source/speech/speech.py [hunk 1]
```diff
@@ -1854,7 +1857,7 @@ def getControlFieldSpeech(  # noqa: C901
 	hasDetailsSequence = getPropertiesSpeech(reason=reason, hasDetails=hasDetails)
 	placeholderSequence = getPropertiesSpeech(reason=reason, placeholder=placeholderValue)
 	nameSequence = getPropertiesSpeech(reason=reason, name=name)
-	valueSequence = getPropertiesSpeech(reason=reason, value=value)
+	valueSequence = getPropertiesSpeech(reason=reason, value=value, _role=role)
```
Comment (seanbudd) [inline source/speech/speech.py:1860]: Why was this change needed?

File: source/IAccessibleHandler/__init__.py [hunk 2]
```diff
@@ -239,6 +244,99 @@
...
+
+def NVDARoleFromAttr(accRole: Union[int, str, None]) -> Role:
+	if accRole.isdigit():
+		accRole = int(accRole)
+	else:
+		accRole = accRole.lower()
+	return IAccessibleRolesToNVDARoles.get(accRole, controlTypes.Role.UNKNOWN)
+
+
+def transformRoleStates(role: Role, states: Set[State]) -> Tuple[Role, Set[State]]:
```
Comment (seanbudd) [inline source/IAccessibleHandler/__init__.py:330]: This function could use some documentation like calculateNvdaStates

File: source/IAccessibleHandler/__init__.py [hunk 3]
```diff
@@ -239,6 +244,99 @@
...
+	# A progress bar that can not convey progress, only activity, is a busy indicator.
+	if roles == Role.PROGRESSBAR and states.intersection({State.HALFCHECKED, State.INDETERMINATE}):
+		states.discard(State.HALFCHECKED)
+		states.add(State.INDETERMINATE)
+		return states
+	return states
+
+
+def NVDARoleFromAttr(accRole: Union[int, str, None]) -> Role:
+	if accRole.isdigit():
```
Comment (seanbudd) [inline source/IAccessibleHandler/__init__.py:323]: `None.isdigit()` and `int.isdigit()` raises an AttributeError

File: source/IAccessibleHandler/__init__.py [hunk 4]
```diff
@@ -5,6 +5,8 @@
 
 # F401 imported but unused. RelationType should be exposed from IAccessibleHandler, in future __all__
 # should be used to export it.
+import typing
+
```
Comment (seanbudd) [inline source/IAccessibleHandler/__init__.py:9]: this has stolen the documentation from the import below it

**Review by seanbudd — COMMENTED** [review]
Hi @feerrenrut, 

I've added a couple of minor notes but otherwise LGTM. 
Would you like a secondary review for the c++ code? It's a fairly straightforward change in my opinion.

Also, in case you missed it - a unit test needs fixing: https://ci.appveyor.com/project/NVAccess/nvda/builds/42786502/tests
This test exists to ensure each state/role has a display string, but I see that this is unnecessary for `State.INDETERMINATE`?

File: source/speech/speech.py (hunk 1 above)
Comment (feerrenrut) [reply source/speech/speech.py:1860]: Role needs to be considered in order to know that value must not be spoken. Passing in `role=role` would result in the role text being added to the output value Sequence. There is already support in this function for underscore prefixed role `_role` to mean consider it, but don't add to output.

File: source/IAccessibleHandler/__init__.py (hunk 3 above)
Comment (feerrenrut) [reply source/IAccessibleHandler/__init__.py:323]: Note that this code is copied verbatim from usages. The type information is fuzzy, it was guessed based on usage.

**Comment by feerrenrut** [comment]
@CyrilleB79 These kinds of progress bars do not convey progress visually and they can be presented in lots of different ways. 

>From a user point of view, busy indicator may let think to the unrelated following concepts:
>
>The browser indicates "busy" when the page is loading
>"busy" may also refer to a mouse pointer shape (e.g. hourglass, clock, etc.) indicating that the computer is busy.

These are related concepts. The main intent of an indeterminate progress bar is to reassure the user that something is happening, even though the amount of work / time remaining is unknown.

>Give the progress bar with undetermined stat a state or value of "running", "working", etc.? I agree that it's not so easy to find a correct wording for it. But creating a new role instead seems more confusing to me.

The main argument against is an indeterminate progress bar widgets do not convey progress, they have no 'value' attribute, so they can't. Announcing "progress bar  working" is longer than it needs to be, the role is misleading and the state is redundant. Essentially, an indeterminate progress bar is a mono-state control.

**Comment by CyrilleB79** [comment]
> @CyrilleB79 These kinds of progress bars do not convey progress visually and they can be presented in lots of different ways.

OK Thanks. Since you mention a great variety of presentations, the busy indicator makes sense.

My last remaining concern is the situation where a progress bar transitions from half-checked (undeterminate) state to (determinate) value = 100%. That is, when a task is in progress without any estimation of the advancement and then is completed.
I have no real situation where this occurs.
* Can this situation occur? If yes, is it common?
* In such case, would NVDA report the role change?

Should this situation be common, I think that the role change may be misleading from a user point of view.
If it is not (what is more likely), there is no need to change the design for such a corner case.

**Comment by feerrenrut** [comment]
>Can this situation occur?

Yes, it could. This, would be author intent though, there may be valid reasons to want to do this.
I have updated the [sample](https://codepen.io/reefturner/pen/KKyoYJb) with this case.

>If yes, is it common?

I doubt it. It's certainly an edge case.

> In such case, would NVDA report the role change?

Currently, NVDA will report the final beep and "100%" (depending on the user's setting for presentation, progress bar output).
With my current test case approach, Chrome seems to swap the IAccessible object when changing from indeterminate to complete determinate. This is a real corner case (author wants to swap from indeterminate to determinate, user wants to monitor the widget for completion), that I suspect there are work arounds for. If it came to it, we could probably resolve the issues (perhaps requiring changes in chrome) but certainly it shouldn't block this initial implementation.

File: source/IAccessibleHandler/__init__.py (hunk 2 above)
Comment (feerrenrut) [reply source/IAccessibleHandler/__init__.py:330]: While writing docs here, I don't think this belongs in `IAccessibleHandler`, it is not specific to IAccessible because it is "post-conversion" and dealing only with NVDA Role and NVDA States.

I'm going to move it.

File: source/IAccessibleHandler/__init__.py (hunk 3 above)
Comment (feerrenrut) [reply source/IAccessibleHandler/__init__.py]: I've fixed this by making it accept an optional string, handle None / empty string explicitly, and assert the type is string. There is one usage that was potentially calling with an `int` value `0`. This will now call with `None`.

File: source/IAccessibleHandler/__init__.py [hunk 5]
```diff
@@ -61,12 +63,14 @@
 NAVRELATION_NODE_CHILD_OF = 0x1005
 NAVRELATION_EMBEDS = 0x1009
 
+if typing.TYPE_CHECKING:
+	import textInfos
```
Comment (lukaszgo1) [inline source/IAccessibleHandler/__init__.py:67]: Any particular reason why this is not placed together with other imports at the top of file?

Comment (feerrenrut) [reply source/IAccessibleHandler/__init__.py:67]: No, I'll move it.

**Review by seanbudd — APPROVED** [review]
Thanks @feerrenrut
